keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,523 @@
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Paths
+
+The overarching `/openapi/v3` endpoint will contain the list of paths (groups) available and serve as a discovery endpoint. Clients can then choose the group(s) to fetch and send additional requests.
+
+/openapi/v3
+
+```json
+{
+   "Paths" : [
+      "api",
+      "api/v1",
+      "apis",
+      "apis/admissionregistration.k8s.io",
+      "apis/apiextensions.k8s.io",
+      "apis/apps",
+      "apis/authentication.k8s.io",
+      ...
+   ]
+}
+```
+
+Based on the provided group, clients can then request `openapi/v3/apis/apps/v1`, `/openapi/v3/apis/networking.k8s.io/v1` and etc. These leaf node specs are self contained OpenAPI v3 specs and include all referenced types.
+
+### Controllers
+
+#### OpenAPI Builder
+
+The OpenAPI builder for CRDs converts the v3 validation schema into a v2 schema. We will keep the conversion to the v2 schema for backwards compatibility, but also generate the v3 schema to publish at the new v3 endpoint.
+
+#### Proto Models & ETags
+
+Kubernetes publishes the openapi schema in both JSON and Protobuf. v3 schemas will include the same mechanisms as v2, publishing a protobuf model of the schema, as well as the corresponding etags for caching. In order to publish the schema based on groups, the CRDs will be grouped by group and published only at the endpoint for their specific group.
+
+#### Aggregator
+
+The aggregator has a mapping of all the APIServices and refreshes the aggregated spec on an interval. APIService already publish by group-version so their behavior is unchanged. Instead of aggregating in the aggregator, we will simply copy the spec to be published at the corresponding aggregator endpoint. For CRDs, instead of downloading the entire spec for CRDs, they will be downloaded per group-version, increasing the number of requests sent internally when a CRD with multiple groups is registered.
+
+### OpenAPI
+
+Swagger 2.0 and OpenAPI 3.0 have a couple of incompatible changes that are not currently supported by kube-openapi. Definitions are moved to components and the schema for Paths is slightly modified. A new struct and builder will be created in kube-openapi to represent the v3 schema to be imported by the necessary consumers.
+
+CRD structural schemas will be published as is without any value stripping for v3. Built-in types will generate the v3 OpenAPI definition, and either keep all fields or strip incompatible fields when building the swagger spec for v3 and v2 respectively.
```

Comment:
please verify that the proto based spec has the same understanding of nullable. Last time I checked it was different, i.e. part of the type.

Reply:
Just looking at the gnostic library for v3 support, I don't see any problems with nullable representation. It's marked as a [bool in the proto](https://github.com/google/gnostic/blob/master/openapiv3/OpenAPIv3.proto#L538) and is handled in a [similar way as other fields](https://github.com/google/gnostic/blob/master/openapiv3/OpenAPIv3.go#L3858-L3866)